src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1291,13 +1309,30 @@ void ObjectSynchronizer::inflate_helper(oop obj) {
 
 // Can be called from non JavaThreads (e.g., VMThread) for FastHashCode
 // calculations as part of JVM/TI tagging.
-static bool is_lock_owned(Thread* thread, oop obj) {
+static bool is_lock_owned(JavaThread* locking_thread, oop obj) {
   assert(LockingMode == LM_LIGHTWEIGHT, "only call this with new lightweight locking enabled");
-  return thread->is_Java_thread() ? JavaThread::cast(thread)->lock_stack().contains(obj) : false;
+  return locking_thread != nullptr ? locking_thread->lock_stack().contains(obj) : false;
 }
 
 ObjectMonitor* ObjectSynchronizer::inflate(Thread* current, oop object,
                                            const InflateCause cause) {
+  JavaThread* locking_thread = nullptr;
+  if (LockingMode == LM_LIGHTWEIGHT && current->is_Java_thread()) {
+    locking_thread = JavaThread::cast(current);
+  }
+  return inflate(locking_thread, current, object, cause);
+}
+
+ObjectMonitor* ObjectSynchronizer::inflate(JavaThread* locking_thread, Thread* current, oop object,
+                                           const InflateCause cause) {
+  assert(current == Thread::current(), "must be");
+  // locking_thread != nullptr implies that the thread is safe to look into including
+  // reading and modifying it's lock stack. This is always valid if current == locking_thread
+  // When called with locking_thread != nullptr and locking_thread != current some mechanism
+  // must synchronize the locking_thread with respect to the current thread. Currently only
+  // used when deoptimizing and re-locking locks. See Deoptimization::relock_objects
+  assert(locking_thread == nullptr || locking_thread == current ||
+         locking_thread->is_obj_deopt_suspend(), "must be");
```

Comment:
Reversing the args here would help.  You could also add that JVMTI gets hash codes in a safepoint that might also inflate the monitor, without any locking thread.